### PR TITLE
feat(conformance): era-awareness so modern servers stop false-failing (Phase 3 §11.5)

### PR DESCRIPTION
**Pull request by chelojimenez**
Phase 3 §11.5. Teaches the MCP conformance suite about the **protocol era** so a conforming modern (2026-07-28) server stops false-failing on checks that only assert 2025-era wire mechanics. Scope is deliberately narrow — **no new modern MUST checks** (that is Phase 7). Legacy runs stay byte-identical.

## What changed

- **`CHECK_ERAS`** (new exported map in `types.ts`) is the single source of truth for which era each check applies to, consumed by BOTH the client-backed track (`eraGate` in the runner) and the raw-HTTP track (protocol/security/transport runners). Phase 7 reuses it.
- **`protocolVersion`** added to `MCPConformanceConfig`/normalized config (additive, optional). Validated via `isKnownProtocolVersion` (throws on unknown); `era` derived via `isStatelessProtocolVersion`. **Unset ⇒ legacy ⇒ byte-identical behavior.**
- Runner now passes `mcpProtocolVersion` into the server config (this is the line that lets a modern-only server connect via version negotiation), era-gates client checks as **skips** (never filtered out, so they still appear in the report), and reads the client through `getManagedClient()` — the dead stateless-preview `getClient()` skip block is gone (the preview client was removed in Phase 1C; `OfficialSdkClientAdapter` is the sole `ManagedMcpClient`).
- `complete()` added to `ManagedMcpClient` + adapter (additive) for the completion check; `logging-set-level` adjusted to the managed `void` return.
- Raw checks: the four hardcoded `"2025-11-25"` sites now use `config.protocolVersion ?? "2025-11-25"`; `protocol-invalid-method-error` gets a **sessionless modern probe** (drops the initialize prelude; sends the bad-method POST with the modern `_meta` protocol-version envelope + `MCP-Protocol-Version`/`mcp-method` headers).

## Decisions to review

- **CHECK_ERAS downgrade (safety valve):** the two `localhost-host-*` security checks are classified **legacy-only**, not both-era as the plan's first-pass classification suggested. Their raw modern host-header probe cannot be made reliably green against the dual-era fixture, which leaves `enableDnsRebindingProtection` **disabled** (verified in the server transport) — so `localhost-host-rebinding-rejected` cannot pass there in *any* era. Per the plan's explicit safety valve ("a skip is always safe; a false failure never is") they surface as era-skips on a modern run. Promoting them to both-era is Phase 7 work with a protecting fixture. All other checks match the plan's classification; `ping` is left legacy-only (not promoted) since it could not be verified on a 2026 pin.
- **`protocol-invalid-method-error` kept both-era** with a modern probe (canonical `METHOD_NOT_FOUND`). This is the one modern raw probe retained.
- **Verification gap (environmental):** the fixture-backed `era.integration.test.ts` (the exit gate) and every existing conformance runner/mock test **could not be executed in the overnight worktree** — `@modelcontextprotocol/server`/`node@2.0.0-beta.4` are not installed and the disk was full, so `npm install` fails (ENOSPC). What WAS verified locally: `sdk` `tsc --noEmit` clean; `era.unit.test.ts` (12 tests, incl. the modern probe framing via a stub fetch) + updated `checks.test.ts` green (17 total). The integration test type-checks (only the pre-existing missing-package error, identical to the existing fixture test). **The modern-run exit-gate assertions should be confirmed by CI**, where the beta.4 packages are present.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches conformance verdict logic, protocol negotiation pins, and raw HTTP probes across legacy and modern eras; regressions could hide real failures or change default legacy runs if era mapping is wrong.
> 
> **Overview**
> Adds **protocol-era awareness** to the MCP conformance suite so pinning **2026-07-28** (modern/sessionless) skips checks that only assert 2025 stateful wire behavior instead of false-failing. Unset `protocolVersion` stays **legacy** and is intended to match prior behavior.
> 
> **`CHECK_ERAS`** maps each check to `legacy` / `modern` and drives **`eraGate`** on client checks plus raw HTTP runners (protocol, security, transport). Optional **`protocolVersion`** on config is validated and normalized to **`era`**; the runner passes **`mcpProtocolVersion`** into the ephemeral client for negotiation.
> 
> Client track uses **`getManagedClient()`** (drops the old stateless-preview skip path). **`complete()`** is added on **`ManagedMcpClient`** / adapter; **`logging-set-level`** passes on resolve only (void). Raw probes use the pinned version; **`protocol-invalid-method-error`** adds a sessionless modern POST (full `_meta`, headers via **`Headers.set`**) and requires JSON-RPC **-32601**. Connect failures on modern runs era-skip inapplicable checks and anchor failure so all-legacy-only selections cannot report **passed** without connecting.
> 
> Tests: **`era.unit.test.ts`**, **`era.integration.test.ts`**, runner edge case.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e66f8e7424137f88df1d8ae8591e688b217c6e50. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
## Summary by cubic
Makes the MCP conformance suite era-aware so modern (2026) runs skip 2025-only checks instead of false-failing. Legacy runs remain unchanged; the invalid-method probe is hardened and connect failures are handled correctly.

- **New Features**
  - Added optional `protocolVersion` to config; derive `era` and validate via `isKnownProtocolVersion`.
  - Introduced `CHECK_ERAS` used by client and raw-HTTP tracks; inapplicable checks are reported as skipped with a consistent message.
  - Runner passes `mcpProtocolVersion`, uses `getManagedClient`, and era-gates client checks.
  - Added `complete()` to `ManagedMcpClient` and `OfficialSdkClientAdapter`; `logging-set-level` now resolves to `void`.
  - Modernized `protocol-invalid-method-error`: sessionless 2026 probe with full `_meta` envelope, requires JSON-RPC `-32601`, and honors `config.protocolVersion ?? "2025-11-25"`.
  - Classified `localhost-host-*` as legacy-only; they skip on modern runs.

- **Bug Fixes**
  - Applied era gating to connect-error paths and, when all selected checks are legacy-only on a modern run, anchored the connect failure to the first check so the run doesn’t report passed; added a regression test that asserts this exact anchor.
  - Built modern invalid-method probe headers with `Headers.set` to dedupe case-variant collisions with `customHeaders` so pins can’t be overridden.

<sup>Written for commit e66f8e7424137f88df1d8ae8591e688b217c6e50. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/MCPJam/inspector/pull/3400?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by chelojimenez** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/marcelo-apk6cgUZZztPZRjPdWkYwH/pr-checks/6246bcbd-b119-4680-b737-6f4f9d1ff1a8?refs=c7097e7903c740326eb401d94553d0dd8d3acf3b&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

**Comment by chelojimenez** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/marcelo-apk6cgUZZztPZRjPdWkYwH/pr-checks/06d7dfad-7488-497d-b364-22ae4fb81c47?refs=e66f8e7424137f88df1d8ae8591e688b217c6e50&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: sdk/src/mcp-conformance/runner.ts [hunk 1]
```diff
@@ -223,6 +236,15 @@ async function runClientChecks(
...
         for (const check of selectedClientChecks) {
+          // Era gate first: an era-inapplicable check is a deterministic skip
+          // that never runs `check.run`, so it neither observes nor affects
+          // the `connectionLost` sequencing below.
+          const eraSkip = eraGate(check, config.era, config.protocolVersion);
+          if (eraSkip) {
+            results.push(eraSkip);
+            continue;
+          }
+
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/runner.ts:247]: Fixed in 2163925f26. The connect-error `catch` path in `runClientChecks` now applies `eraGate` first: an era-inapplicable check (e.g. `server-initialize` on a modern run) is emitted as a deterministic era-skip instead of a failure, so it can no longer drag `result.passed` false. The surfaced connect failure is pinned to the first *era-applicable* check (in the legacy era that is still `selectedClientChecks[0]`, byte-identical to before).

File: sdk/src/mcp-conformance/runner.ts [hunk 2]
```diff
@@ -223,6 +236,15 @@ async function runClientChecks(
         let connectionLost = false;
 
         for (const check of selectedClientChecks) {
+          // Era gate first: an era-inapplicable check is a deterministic skip
+          // that never runs `check.run`, so it neither observes nor affects
+          // the `connectionLost` sequencing below.
+          const eraSkip = eraGate(check, config.era, config.protocolVersion);
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/runner.ts:242]: Fixed in 2163925f26 — same change as the sibling thread. `eraGate` is now applied in the connect-error `catch` path before the failed/skipped partition, and the connect failure is pinned to the first era-applicable check. Legacy runs are byte-identical (gate is inert). Verified against era.integration/era.unit/checks/runner suites (23/23 pass).

File: sdk/src/mcp-conformance/checks/protocol.ts [hunk 3]
```diff
@@ -58,35 +66,74 @@ async function initializeAndGetSession(
...
         jsonrpc: "2.0",
         id: 99,
-        method: "nonexistent/method_that_does_not_exist",
-        params: {},
+        method: INVALID_METHOD,
+        params: {
+          _meta: {
+            "io.modelcontextprotocol/protocolVersion": version,
+          },
+        },
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/checks/protocol.ts:129]: Fixed in 2163925f26. In the modern probe `customHeaders` (via `buildBaseHeaders`) is now spread BEFORE the probe-owned `MCP-Protocol-Version` / `mcp-method` headers, so a caller header can never overwrite the protocol pin or the SEP-2243 method mirror.

File: sdk/src/mcp-conformance/checks/protocol.ts [hunk 4]
```diff
@@ -58,35 +66,74 @@ async function initializeAndGetSession(
...
     });
+  }
+
+  const sessionId = await initializeAndGetSession(ctx);
+  return await ctx.fetchFn(ctx.serverUrl, {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+      Accept: "application/json, text/event-stream",
+      ...buildBaseHeaders(ctx),
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/checks/protocol.ts:140]: Fixed in 2163925f26. Base headers (which carry `customHeaders`) are now spread before the probe-owned `MCP-Protocol-Version` and `mcp-method` headers, keeping the version, method header, and `_meta` value aligned regardless of caller headers.

File: sdk/src/mcp-conformance/checks/protocol.ts [hunk 5]
```diff
@@ -58,35 +66,74 @@ async function initializeAndGetSession(
...
+): Promise<MCPCheckResult[]> {
+  const results: MCPCheckResult[] = [];
+
+  if (!selectedCheckIds.has("protocol-invalid-method-error")) {
+    return results;
+  }
+
+  const startedAt = Date.now();
+  try {
+    const response = await sendInvalidMethodProbe(ctx);
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/checks/protocol.ts:180]: Fixed in 2163925f26. The probe now requires `rpcError.code === -32601` (Method not found) and fails otherwise, so a server returning an unrelated JSON-RPC error code is no longer a false pass. Coupled with the modern-envelope fix on the sibling thread (the dual-era fixture confirmed a minimal envelope was being rejected with -32602 before dispatch), the check now genuinely exercises unknown-method handling in both eras.

File: sdk/src/mcp-conformance/checks/protocol.ts [hunk 6]
```diff
@@ -58,35 +66,74 @@ async function initializeAndGetSession(
...
       body: JSON.stringify({
         jsonrpc: "2.0",
         id: 99,
-        method: "nonexistent/method_that_does_not_exist",
-        params: {},
+        method: INVALID_METHOD,
+        params: {
+          _meta: {
+            "io.modelcontextprotocol/protocolVersion": version,
+          },
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/checks/protocol.ts:128]: Fixed in 2163925f26 (not deferred). Confirmed real against the dual-era fixture, which runs the actual 2026-07-28 `_meta` envelope validation and rejected the minimal probe with: `Invalid _meta envelope for protocol revision 2026-07-28: io.modelcontextprotocol/clientInfo: missing` (-32602) — before ever dispatching the bogus method. The modern probe now sends the full envelope (`io.modelcontextprotocol/protocolVersion` + `clientInfo` + `clientCapabilities`, mirroring the legacy `initialize`), so a conforming server reaches method dispatch and returns -32601. Combined with the new -32601 assertion, the check no longer false-passes on an envelope-rejection error. We completed the envelope rather than using the legacy-only safety valve because the PR's tests intentionally pin this check as both-era, and the fixture gives an exact, validated envelope contract (so it is not fragile guesswork).

[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/checks/protocol.ts:128]: Fixed in 2163925f26. Added both `io.modelcontextprotocol/clientInfo` and `io.modelcontextprotocol/clientCapabilities: {}` to the modern probe's `_meta` (the SDK's 2026-07-28 envelope requires all three keys; the dual-era fixture rejected the incomplete envelope with -32602 before dispatch). The modern probe now reaches unknown-method handling and returns -32601. A unit assertion pins all three envelope keys.

File: sdk/src/mcp-conformance/types.ts [hunk 7]
```diff
@@ -35,6 +38,59 @@ export type MCPCheckId = (typeof MCP_CHECK_IDS)[number];
...
+ *     (`tools-list`, `tools-input-schemas-valid`, `prompts-list`,
+ *     `resources-list`, `protocol-invalid-method-error`).
+ *
+ * The two `localhost-host-*` security checks are deliberately legacy-only for
+ * now: their raw modern host-header probe could not be validated against the
+ * dual-era fixture (which leaves DNS-rebinding protection disabled), so per
+ * the Phase 3 safety valve they are downgraded to a safe skip on a modern run
+ * rather than shipped as a fragile probe. Promoting them is Phase 7 work.
+ */
+export const CHECK_ERAS: Record<MCPCheckId, readonly MCPCheckEra[]> = {
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/types.ts:76]: Fixed in 2163925f26. `runProtocolChecks` now consults `CHECK_ERAS` and emits an era-skip when the check does not apply to the run's era, mirroring the security and transport raw runners. `protocol-invalid-method-error` remains classified for both eras, so this is inert today — but the raw protocol track is now on the same single-source-of-truth map, so a future reclassification becomes a safe era-skip instead of silently drifting into a false failure.

File: sdk/src/mcp-conformance/runner.ts [hunk 8]
```diff
@@ -259,11 +281,30 @@ async function runClientChecks(
...
+      const eraSkip = eraGate(check, config.era, config.protocolVersion);
+      if (eraSkip) {
+        checks.push(eraSkip);
+        continue;
+      }
+
+      if (
+        check.id === "server-initialize" ||
+        check.id === firstApplicableCheck?.id
+      ) {
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/runner.ts:328]: Fixed in e25e2a5983. On a modern run where every selected client check is legacy-only, `firstApplicableCheck` was undefined and the connect failure got era-skipped away — the run could report `passed`. The catch path now anchors the connect failure on the first selected check when no era-applicable check exists, so a genuine connect failure always surfaces as >=1 failed result. The branch is guarded by `!firstApplicableCheck`, which is unreachable in the legacy era (every check applies to `legacy`), so legacy runs stay byte-identical. New test in `runner.test.ts`: modern + connect failure + `checkIds: ["ping"]` (legacy-only) now asserts `passed === false`.

File: sdk/src/mcp-conformance/runner.ts [hunk 9]
```diff
@@ -259,11 +281,30 @@ async function runClientChecks(
...
     const checks: MCPCheckResult[] = [];
 
+    // Era gate applies even when the connection could not be established: an
+    // era-inapplicable check is a deterministic era-skip, never a failure. On a
+    // modern run a connect error must NOT surface a legacy-only check (e.g.
+    // `server-initialize`) as failed, or `result.passed` goes false for a check
+    // that does not apply to the run's era. The surfaced connect failure is
+    // pinned to the first ERA-APPLICABLE check (in legacy that is
+    // `selectedClientChecks[0]`, byte-identical to the pre-era-awareness path).
+    const firstApplicableCheck = selectedClientChecks.find(
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/runner.ts:293]: Fixed in e25e2a5983 (same root cause as the cursor report). The connect-error catch path now anchors the failure on the first selected check when `firstApplicableCheck` is undefined (all-era-skipped modern run), so the connection error is never silently discarded. Guarded so the legacy path is byte-identical. Covered by a new all-legacy-only + connect-failure test in `runner.test.ts`.

File: sdk/src/mcp-conformance/checks/protocol.ts [hunk 10]
```diff
@@ -58,35 +74,105 @@ async function initializeAndGetSession(
...
         Accept: "application/json, text/event-stream",
         ...buildBaseHeaders(ctx),
-        ...(sessionId ? { "mcp-session-id": sessionId } : {}),
+        // Probe-owned headers spread LAST so a caller's `customHeaders` can
+        // never overwrite the protocol-version pin or the SEP-2243 method
+        // mirror — otherwise a stray custom header turns an honest probe into
+        // a false failure unrelated to the server under test.
+        "MCP-Protocol-Version": version,
+        "mcp-method": INVALID_METHOD,
       },
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/checks/protocol.ts]: Valid — fixed in e25e2a5983. The prior fix relied on object-key ordering, but a plain object keeps `MCP-Method` and `mcp-method` as distinct keys; Fetch then normalizes both to canonical `mcp-method` and concatenates the values. The modern probe now builds a `Headers` instance and `headers.set(...)` the `MCP-Protocol-Version` / `mcp-method` pins after the base spread, so a caller `customHeaders` case-variant is replaced (case-insensitive by name) rather than blended. New `era.unit` test feeds `customHeaders: { "MCP-Method": "tools/list", "mcp-protocol-version": "1999-01-01" }` and asserts the probe pins win cleanly with no concatenation. Legacy branch left untouched (no probe-critical pins there, so it stays byte-identical).

File: sdk/src/mcp-conformance/types.ts [hunk 11]
```diff
@@ -58,6 +114,14 @@ export interface MCPConformanceConfig {
...
   fetchFn?: typeof fetch;
   clientName?: string;
+  /**
+   * Pinned MCP protocol version. Absent ⇒ legacy era, byte-identical to the
+   * pre-era-awareness behavior. A known stateless value (e.g. `"2026-07-28"`)
+   * selects the modern era: the client connects through the official Client's
+   * version negotiation and era-scoped checks apply per {@link CHECK_ERAS}.
+   * Validated at normalization via `isKnownProtocolVersion`.
+   */
+  protocolVersion?: McpProtocolVersion;
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/types.ts:124]: Flagging for morning review — deferring to a clean follow-up. Threading `mcpProtocolVersion` from the Inspector conformance ROUTES (`server/routes/mcp/conformance.ts`, `server/routes/web/conformance.ts`, `server/routes/shared/conformance.ts`) into `MCPConformanceConfig.protocolVersion` is inspector-route plumbing, out of scope for this SDK-core era-awareness PR. The SDK field and era gating land here; the product-flow wiring is a self-contained next PR. Not blocking this one.

File: sdk/tests/mcp-conformance/runner.test.ts [hunk 12]
```diff
@@ -68,6 +68,25 @@ describe("MCPConformanceTest", () => {
...
+      protocolVersion: "2026-07-28",
+      checkTimeout: 3_000,
+      checkIds: ["ping"],
+    });
+
+    const result = await test.run();
+
+    expect(result.passed).toBe(false);
+    const failed = result.checks.filter((check) => check.status === "failed");
+    expect(failed.length).toBeGreaterThanOrEqual(1);
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/tests/mcp-conformance/runner.test.ts]: Fixed in e66f8e7 — the test now asserts the anchored check (`ping`) is the one that failed AND that it is the only failure (`failed.map(id) === ["ping"]`), so it can no longer pass on an unrelated failure that masks a missing anchor.

File: sdk/tests/mcp-conformance/runner.test.ts [hunk 13]
```diff
@@ -68,6 +68,25 @@ describe("MCPConformanceTest", () => {
     }
   });
 
+  it("modern run + connect failure + all-legacy-only selection surfaces a failure, not a pass", async () => {
+    // Modern era, but the only selected client check (`ping`) is legacy-only,
+    // so it is era-skipped for this run. The server is unreachable, so
+    // `withEphemeralClient` throws. Without the connect-failure anchor, every
+    // check would be era-skipped and the run would silently report `passed`.
+    const test = new MCPConformanceTest({
+      serverUrl: "http://127.0.0.1:1/mcp",
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/tests/mcp-conformance/runner.test.ts:77]: Addressed alongside e66f8e7: `127.0.0.1:1` is the conventional unreachable test address (port 1 is privileged and effectively never bound in CI). More importantly, the strengthened assertion now requires `ping` specifically to be the sole failure — so even in the unlikely event a service occupied that port, a non-MCP endpoint would still surface the anchored `ping` failure rather than a silent pass. Flagging the theoretical port choice for morning review if you'd prefer an ephemeral-bind-then-close pattern.

File: sdk/src/mcp-conformance/checks/transport.ts [hunk 14]
```diff
@@ -304,14 +308,38 @@ export async function runTransportChecks(
...
+        skippedResult(
+          TRANSPORT_CHECK_METADATA[id],
+          eraSkipMessage(ctx.config.era, ctx.config.protocolVersion),
+        ),
+      );
+    }
+  }
+
+  if (applicableTransportChecks.length === 0) {
+    return results;
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply sdk/src/mcp-conformance/checks/transport.ts:340]: Deferred (design), not fixed — this is distinct from the connect-failure anchor. There, the server WAS contacted and the connection genuinely failed, so hiding it was wrong. Here, a modern run that selects only legacy-only transport checks era-SKIPS them because they do not apply to the modern era — and §11.5's explicit design is 'a modern run reports them N/A with a reason, not a red failure', with skipped-never-fails as an intentional safety valve (a false failure is never acceptable). An all-N/A run reporting passed is the honest 'nothing applicable ran', the same as running zero checks; forcing it red would be exactly the false-failure the design forbids.

The legitimate improvement is a READINESS warning ('0 era-applicable checks ran for this selection'), which is a UX signal rather than a conformance failure — flagging that for @chelojimenez's morning review as an optional enhancement rather than changing the pass/fail semantics here.
